lib/test_queue/runner.rb

```diff
@@ -427,40 +429,55 @@ def distribute_queue
         else
           sock = @server.accept
           cmd = sock.gets.strip
+
+          token = cmd[TOKEN_REGEX, 1]
+          # If we have a slave from a different test run, don't respond, and it will consider the test run done.
+          case token
+          # No special handling if the token is valid
+          when @run_token
+          when nil
+            STDERR.puts "*** Worker sent no token to master for run #{@run_token}; ignoring."
+            sock.write("WRONG RUN\n")
+            File.open("/tmp/errors.log", "a") {|f| f.puts "WRONG RUN - nil token"}
+            next
+          else
+            STDERR.puts "*** Worker from run #{run_token} connected to master for run #{@run_token}; ignoring."
+            sock.write("WRONG RUN\n")
+            File.open("/tmp/errors.log", "a") {|f| f.puts "WRONG RUN - wrong token"}
+            next
+          end
+
+          cmd = cmd.gsub(TOKEN_REGEX, "").strip
+
           case cmd
-          when /^POP/
+          when /POP/
```

Comment:
If `TOKEN_REGEX` also consumed trailing spaces you could keep the `^`. That might be nice in case we add a `ICEYPOP` command later (not exactly :trollface:).